### ISSUE TITLE
linter: react/set-state-in-effect misses setState behind a guard derived from a ref read

### ISSUE THREAD

**Issue by Neaox** [oxc-project/oxc#26007]
### What version of Oxlint are you using?

1.79.0 (oxlint-tsgolint 7.0.2001; type-aware mode not needed to reproduce)

### What command did you run?

`oxlint case.jsx`

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "plugins": ["react"],
  "categories": { "correctness": "off" },
  "rules": {
    "react/set-state-in-effect": "error"
  }
}
```

### What happened?

`react/set-state-in-effect` reports nothing when the effect early-returns on a value derived from a ref read and then calls `setState` unconditionally. `eslint-plugin-react-hooks` reports the same file.

`case.jsx`:

```jsx
import { useEffect, useRef, useState } from "react"

export function Bar({ visible }) {
  const prevVisible = useRef(0)
  const [direction, setDirection] = useState(null)

  useEffect(() => {
    const dv = visible - prevVisible.current
    if (dv === 0) return
    setDirection("right")
  }, [visible])

  return <p>{direction}</p>
}
```

```
$ oxlint case.jsx
Found 0 warnings and 0 errors.
```

Same file under `eslint-plugin-react-hooks@7.0.1` (`react-hooks/set-state-in-effect: error`, eslint 10.9.0):

```
10:5  error  Calling setState synchronously within an effect can trigger cascading renders
       > 10 |     setDirection("right")
            |     ^^^^^^^^^^^^ Avoid calling setState() directly within an effect
```

Variants, each 1 finding under eslint-plugin-react-hooks:

| variant | oxlint |
| --- | --- |
| early return on a prop: `if (visible === 0) return; setDirection(…)` | 1 |
| ref write then unconditional setState: `prevVisible.current = visible; setDirection(…)` | 1 |
| ref read into a local, early return on that local, then setState (case above) | 0 |

Adding the ref write back before the guard, or wrapping the component in `memo()`, does not change the result.

Expected: a diagnostic at `setDirection("right")`, as with eslint-plugin-react-hooks.

Platform: Node 24.19.0, Windows 11 x64.

Disclosure per the AI usage policy: the reduction was done with an AI coding agent; I have re-run the reproduction and reviewed the report.

**Comment by Neaox** [comment]
Re-tested with `eslint-plugin-react-hooks` 7.1.1 (current release): it does not report this case either. The comparison in the original report was against 7.0.1, which did.

| case | oxlint 1.79.0 | react-hooks 7.0.1 | react-hooks 7.1.1 |
| --- | --- | --- | --- |
| `setState` behind a guard derived from a ref read (case above) | 0 | 1 | 0 |
| same guard on a prop (control) | 1 | 1 | 1 |

So oxlint matches current upstream behaviour; this is not a false negative. Closing. Apologies for the noise.
